rfcs/20201113-tf-core.md

```diff
@@ -0,0 +1,175 @@
+# Tensorflow Core as a product RFC
+
+| Status        | Proposed                                            |
+:-------------- |:---------------------------------------------------- |
+| **Authors**   | Yanhua Sun (yanhuasun@google.com) |
+| **Sponsor**   | Rohan Jain (rohanj@google.com)        |
+| **Updated**   | 2020-11-13                                           |
+
+
+## Objective
+
```

Comment:
This needs some more nuance.

When I first read "TF core" I thought it was about separating the C++ library from the python; but below it seems like you want to include some python API as part of core.

So for example when it comes to XLA/MLIR, either JITted or AOT (e.g., the functionality in saved_model_cli), do you plan to keep those in core, or do you expect that all XLA functionality be split off in separate symbolic libraries?  If so, how do you expect to support `jit_compile=True` extension of tf.function?

Reply:
+1 for separating C++ library from the Python one. The Python API might serve for prototyping purpose if we'd like to collect feedbacks from a broader audience though, e.g. https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/distribute/v1/all_reduce.py

Reply:
I'll try and clarify the proposal a bit here. There are couple of directions this RFC is pushing towards

1. Python API modularity - Currently TF core includes Keras, tf.data, distribution strategies etc. without any clear layering and dependency structure. We want to define a small core on which all these libraries can be layered on top of. We don't plan to remove any APIs from TF or add too many new APIs etc. 

2. C++ first API - We want to think of the "Core API" as primarily a C++ one with a thin (probably 1-1 mapping) layer of python on top. Users can choose to interact with either API. Arguably, one can think of this as an implementation detail but its a real important implementation detail that we wanted to bring folks attention to.

Concretely, the first step we're taking now is defining the boundaries of what this small core API surface is going to look like and ensuring that the other TF APIs (tf.data, distribute, keras etc.) are dependent on it (and not the other way). This involves removing circular dependencies, BUILD file refactoring, moving some files around, creating right extension points / interfaces in TF core with no API surface changes expected whatsoever.

Reply:
@rohan100jain +1
If there is design meeting kindly let us know. I would like to join the call.

Reply:
Yes, design review meeting is scheduled at 10AM PT Dec 15, 2020

Reply:
Sorry, we need more time to figure out things better, the design review is cancelled.